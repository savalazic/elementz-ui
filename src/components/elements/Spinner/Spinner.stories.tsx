import React from 'react';

import { storiesOf } from '@storybook/react';

import { infoConfig } from '../../../utils/info-config';

import { Spinner } from './Spinner';

const stories = storiesOf('Elements|Spinner', module);

stories.addParameters(infoConfig);

stories
  .add('default', () => {
    return <Spinner />;
  })
  .add('with custom size', () => {
    return <Spinner size="200px" />;
  })
  .add('with custom color', () => {
    return <Spinner color="#409eff" />;
  })
  .add('with custom stroke width', () => {
    return <Spinner strokeWidth={6} />;
  });
