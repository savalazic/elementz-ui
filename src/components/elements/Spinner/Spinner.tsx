import React from 'react';
import styled, { keyframes } from 'styled-components';

export interface SpinnerProps {
  /** color */
  color?: string;
  /** @default 50px */
  /** size */
  size?: string;
  /** @default 2 */
  /** stroke width */
  strokeWidth?: number;
  /** @default true */
  /** visibility */
  visible?: boolean;
}

export const rotate = keyframes`
  100% {
    transform: rotate(360deg);
  }
`;

export const dash = keyframes`
  0% {
    stroke-dasharray: 1, 200;
    stroke-dashoffset: 0;
  }
  50% {
    stroke-dasharray: 89, 200;
    stroke-dashoffset: -35px;
  }
  100% {
    stroke-dasharray: 89, 200;
    stroke-dashoffset: -124px;
  }
`;

const StyledSpinner = styled.div<SpinnerProps>`
  display: ${props => !props.visible && 'none'};
  position: relative;
  width: ${props => props.size};

  &::before {
    content: '';
    display: block;
    padding-top: 100%;
  }
`;

const StyledCircular = styled.svg`
  height: 100%;
  transform-origin: center center;
  width: 100%;
  position: absolute;
  top: 0;
  bottom: 0;
  left: 0;
  right: 0;
  margin: auto;
  animation: ${rotate} 2000ms linear infinite;
`;

const StyledPath = styled.circle<SpinnerProps>`
  stroke-dasharray: 1, 200;
  stroke-dashoffset: 0;
  stroke: ${props => props.color || props.theme.default};
  stroke-linecap: round;
  stroke-width: ${props => props.strokeWidth};
  animation: ${dash} 1500ms ease-in-out infinite;
`;

export const Spinner = ({
  color,
  size = '50px',
  strokeWidth = 2,
  visible = true,
}: SpinnerProps) => (
  <StyledSpinner size={size} visible={visible}>
    <StyledCircular viewBox="25 25 50 50">
      <StyledPath
        color={color}
        strokeWidth={strokeWidth}
        cx="50"
        cy="50"
        r="20"
        fill="none"
        strokeMiterlimit="10"
      />
    </StyledCircular>
  </StyledSpinner>
);

export default Spinner;
