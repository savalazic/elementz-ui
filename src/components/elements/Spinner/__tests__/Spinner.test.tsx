import 'jest-dom/extend-expect';
import React from 'react';
import { render } from 'react-testing-library';

import { Spinner } from '../Spinner';

describe('Spinner component', () => {
  it('should match default snapshot', () => {
    const { container } = render(<Spinner />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with width custom size', () => {
    const { container } = render(<Spinner size="50px" />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with width custom color', () => {
    const { container } = render(<Spinner color="#333" />);
    expect(container).toMatchSnapshot();
  });

  it('should match snapshot with custom border width', () => {
    const { container } = render(<Spinner borderWidth="6px" />);
    expect(container).toMatchSnapshot();
  });
});
