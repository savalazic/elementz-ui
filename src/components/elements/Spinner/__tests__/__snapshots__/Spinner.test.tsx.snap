// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner component should match default snapshot 1`] = `
<div>
  <div
    class="sc-bdVaJa xgHQi"
  >
    <svg
      class="sc-bwzfXH hPSsPt"
      viewBox="25 25 50 50"
    >
      <circle
        class="sc-htpNat gmVZZL"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
        stroke-width="2"
      />
    </svg>
  </div>
</div>
`;

exports[`Spinner component should match snapshot custom stroke width 1`] = `
<div>
  <div
    class="sc-bdVaJa xgHQi"
  >
    <svg
      class="sc-bwzfXH hPSsPt"
      viewBox="25 25 50 50"
    >
      <circle
        class="sc-htpNat dXYSBZ"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
        stroke-width="4"
      />
    </svg>
  </div>
</div>
`;

exports[`Spinner component should match snapshot width custom color 1`] = `
<div>
  <div
    class="sc-bdVaJa xgHQi"
  >
    <svg
      class="sc-bwzfXH hPSsPt"
      viewBox="25 25 50 50"
    >
      <circle
        class="sc-htpNat hHYMVt"
        color="#333"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
        stroke-width="2"
      />
    </svg>
  </div>
</div>
`;

exports[`Spinner component should match snapshot width custom size 1`] = `
<div>
  <div
    class="sc-bdVaJa xgHQi"
  >
    <svg
      class="sc-bwzfXH hPSsPt"
      viewBox="25 25 50 50"
    >
      <circle
        class="sc-htpNat gmVZZL"
        cx="50"
        cy="50"
        fill="none"
        r="20"
        stroke-miterlimit="10"
        stroke-width="2"
      />
    </svg>
  </div>
</div>
`;
